### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by xdayaan**

### CODE REVIEW

File: v1/app/modules/vc/schemas/vc_schema.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    type: str
+    token: str
+    url: str
+    added_by_user_id: int
+    created_by: str
+    
+    class Config:
+        orm_mode = True
+
+    class Config:
```
Comment (shabangg) [inline v1/app/modules/vc/schemas/vc_schema.py:34]: Duplicate

Comment (xdayaan) [reply v1/app/modules/vc/schemas/vc_schema.py:34]: Fixing and creating PR again.

File: v1/app/utils/get_repos.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import httpx
+from typing import List
+from app.utils.encodeBase64 import encode_basic_token
+
+async def get_repos(url: str, token: str, git: str) -> List[dict]:
+    print("url:", url)
```
Comment (shabangg) [inline v1/app/utils/get_repos.py:6]: Remove print add logger

File: v1/app/modules/vc/schemas/vc_schema.py (hunk 1 above)
Comment (xdayaan) [reply v1/app/modules/vc/schemas/vc_schema.py]: Fixed the duplicate issue

File: v1/app/utils/get_repos.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    print("token:", token)
+    print("git", git)
+    all_repos = []
+    page = 1
+    per_page = 10
```
Comment (shabangg) [inline v1/app/utils/get_repos.py:11]: Increase this to 50 for speed

File: v1/app/modules/repository/repository_service.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+            # Store the fetched repository data in the database
+            for repo_data in repos_data:
+                repo = Repo(
+                    vc_id=vc.id,
+                    vctype=vc.type,
+                    name=repo_data.get('name'),
+                    repoUrl=repo_data.get('links', {}).get('html', {}).get('href'),
+                    author=repo_data.get('owner', {}).get('display_name'),
+                    other_repo_details="details"
```
Comment (shabangg) [inline v1/app/modules/repository/repository_service.py:38]: Shouldnt this be repo_data.details. We should save all other details as well.

File: v1/app/modules/repository/respository_controller.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        )
+        return {"message": "Repositories fetched and stored successfully."}
+    except Exception as e:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail=str(e)
+        )
+
+@router.get("/")
+async def get_repos(vc_id: int, repo_name: Optional[str] = None, db: AsyncSession = Depends(get_db)):
```
Comment (shabangg) [inline v1/app/modules/repository/respository_controller.py:27]: Add sort_by(last_scan_time, created_at) and order_by(asc/desc)

File: v1/app/modules/repository/models/repository.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, JSON, Enum
+from sqlalchemy.orm import relationship
+from datetime import datetime, timezone
+from app.core.db import Base
+from app.modules.vc.models.vc import VcTypes  # Import VcTypes enum
+
+class Repo(Base):
```
Comment (shabangg) [inline v1/app/modules/repository/models/repository.py:7]: Add created_at

File: v1/app/main.py [hunk 7]
```diff
@@ -1,32 +1,39 @@
...
-            break 
+            admin_user = await get_user_by_username(db, "admin")
+            if not admin_user:
+                admin_user_data = UserCreate(
+                    username="admin",
+                    password=get_password_hash("admin"),
+                    role="admin",
+                    active=True
+                )
+                await create_user(db, admin_user_data)
```
Comment (shabangg) [inline v1/app/main.py:35]: Update this to be `await create_user(db, admin_user_data, User())`

File: v1/app/modules/vc/vc_controller.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+@router.post("/", response_model=VCResponse, status_code=status.HTTP_201_CREATED)
+def create_vc(vc: VCCreate, db: Session = Depends(get_db)):
+    return VCService.create_vc(db=db, vc=vc)
+
+@router.get("/{vc_id}", response_model=VCResponse)
+def get_vc(vc_id: int, db: Session = Depends(get_db)):
+    return VCService.get_vc(db=db, vc_id=vc_id)
+
+@router.get("/", response_model=List[VCResponse])
+def get_all_vcs(vc_type: Optional[str] = None, db: Session = Depends(get_db)):
```
Comment (shabangg) [inline v1/app/modules/vc/vc_controller.py:19]: Add sort_by(vc_type, created_at ) and order_by(asc, desc)

File: v1/app/modules/vc/vc_controller.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+from app.modules.vc.schemas.vc_schema import VCCreate, VCResponse
+from app.core.db import get_db
+from app.modules.vc.vc_service import VCService
+
+router = APIRouter(prefix="/vc")
+
+@router.post("/", response_model=VCResponse, status_code=status.HTTP_201_CREATED)
+def create_vc(vc: VCCreate, db: Session = Depends(get_db)):
+    return VCService.create_vc(db=db, vc=vc)
+
```
Comment (shabangg) [inline v1/app/modules/vc/vc_controller.py:13]: Add filter api -> should return string[] -> [vc_type]
Add filter value api -> should accept a return from `distinct(key) from table`

File: v1/app/modules/repository/respository_controller.py (hunk 5 above)
Comment (shabangg) [reply v1/app/modules/repository/respository_controller.py:27]: Add vc_type filter

File: v1/app/modules/webhookConfig/schemas/webhook_schema.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    loose = "Loose"
+    aggressive = "Aggressive"
+
+class WebhookConfigBase(BaseModel):
+    vc_id: int
+    vc_type: VcTypes
+    scan_type: Optional[AllowedScanType] = AllowedScanType.loose
+    pr_comment: Optional[str] = "Secrets Caught!"
+    block_pr: Optional[bool] = False
+    git_actions: Optional[List[str]] = None
```
[the remark this answers is not in the document]
Comment (xdayaan) [reply v1/app/modules/webhookConfig/schemas/webhook_schema.py:20]: The git_actions and pr_comment?

[the remark this answers is not in the document]
Comment (shabangg) [reply v1/app/modules/webhookConfig/schemas/webhook_schema.py:20]: yes

File: v1/app/utils/get_repos.py (hunk 2 above)
Comment (xdayaan) [reply v1/app/utils/get_repos.py]: done

File: v1/app/utils/get_repos.py (hunk 3 above)
Comment (xdayaan) [reply v1/app/utils/get_repos.py:8]: done

File: v1/app/modules/repository/repository_service.py (hunk 4 above)
Comment (xdayaan) [reply v1/app/modules/repository/repository_service.py:37]: done

File: v1/app/modules/repository/respository_controller.py (hunk 5 above)
Comment (xdayaan) [reply v1/app/modules/repository/respository_controller.py:27]: done

File: v1/app/modules/repository/models/repository.py (hunk 6 above)
Comment (xdayaan) [reply v1/app/modules/repository/models/repository.py:7]: done

File: v1/app/main.py (hunk 7 above)
Comment (xdayaan) [reply v1/app/main.py:35]: done

File: v1/app/modules/vc/vc_controller.py (hunk 9 above)
Comment (xdayaan) [reply v1/app/modules/vc/vc_controller.py:13]: is this the required response?

[
    "bitbucket",
    "github",
    "gitlab"
]

File: v1/app/modules/webhookConfig/schemas/webhook_schema.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (xdayaan) [reply v1/app/modules/webhookConfig/schemas/webhook_schema.py:20]: This is what I came up with:

`from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, JSON, Enum, Boolean
from sqlalchemy.orm import relationship
from datetime import datetime, timezone
from sqlalchemy.ext.declarative import declarative_base
from enum import Enum as PyEnum

Base = declarative_base()

# Enums for Allowed Scan Type
class AllowedScanType(PyEnum):
    loose = "loose"
    aggressive = "aggressive"

# Enums for PR Comments
class PRComment(PyEnum):
    secrets_caught = "Secrets Caught!"
    # Add more comments if needed

# Enums for GitHub Actions
class GitHubActions(PyEnum):
    opened = "opened"
    closed = "closed"
    merged = "merged"
    reopened = "reopened"
    edited = "edited"
    synchronize = "synchronize"

# Enums for GitLab Actions
class GitLabActions(PyEnum):
    open = "open"
    close = "close"
    merge = "merge"
    reopen = "reopen"
    update = "update"

# Enums for Bitbucket Actions
class BitbucketActions(PyEnum):
    opened = "opened"
    declined = "declined"
    merged = "merged"
    reopened = "reopened"
    updated = "updated"

class Repo(Base):
    __tablename__ = 'repositories'

    id = Column(Integer, primary_key=True, index=True)
    vc_id = Column(Integer, ForeignKey('vcs.id'), nullable=False, index=True)
    vctype = Column(Enum('VcTypes', values=['GitHub', 'GitLab', 'Bitbucket']), nullable=False, index=True)
    name = Column(String, index=True)
    repoUrl = Column(String, nullable=False)
    author = Column(String, nullable=False)
    other_repo_details = Column(JSON)
    lastScanDate = Column(DateTime, default=lambda: datetime.now(tz=timezone.utc))
    created_at = Column(DateTime, default=lambda: datetime.now(tz=timezone.utc))

    # Enums
    scan_type = Column(Enum(AllowedScanType), default=AllowedScanType.loose)
    pr_comment = Column(Enum(PRComment), default=PRComment.secrets_caught)
    block_pr = Column(Boolean, default=False)

    # VC Type and Actions
    vc_type = Column(String, nullable=False)  # Store VC type like 'GitHub', 'GitLab', etc.
    git_actions = Column(JSON)  # Store actions as a JSON list of strings

    # Relationships
    vc = relationship('VC', back_populates='repositories')

    def update_git_actions(self):
        if self.vc_type == "GitHub":
            self.git_actions = [action.value for action in GitHubActions]
        elif self.vc_type == "GitLab":
            self.git_actions = [action.value for action in GitLabActions]
        elif self.vc_type == "Bitbucket":
            self.git_actions = [action.value for action in BitbucketActions]
        else:
            self.git_actions = []

`

File: v1/app/modules/vc/vc_controller.py (hunk 9 above)
Comment (shabangg) [reply v1/app/modules/vc/vc_controller.py:13]: Yes

File: v1/app/modules/webhookConfig/schemas/webhook_schema.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (shabangg) [reply v1/app/modules/webhookConfig/schemas/webhook_schema.py:20]: We can just create an enum with all and validation we can do on service. 
Can restrict on FE also. Just create a map { [vc_type]: [allowed_git_actions] }
Eg - 
Enum = all actions
Map = 
```
{ "gitlab" : [ "open", "close" ... ],
"github": [...],
"bitbucket": [...]
}
```

File: v1/app/modules/vc/vc_controller.py (hunk 8 above)
Comment (xdayaan) [reply v1/app/modules/vc/vc_controller.py]: Done

File: v1/app/modules/vc/vc_controller.py (hunk 9 above)
Comment (xdayaan) [reply v1/app/modules/vc/vc_controller.py:13]: Done

File: v1/app/modules/webhookConfig/schemas/webhook_schema.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (xdayaan) [reply v1/app/modules/webhookConfig/schemas/webhook_schema.py:20]: Got it

**Pull request merged**